### ISSUE TITLE
Ansible winrm 'SSL23_GET_SERVER_HELLO:unknown protocol' on non-ssl port

### ISSUE THREAD

**Issue by bence-takacs** [ansible/ansible#16520]
- Bug Report
##### ANSIBLE VERSION

ansible 1.9.1
##### CONFIGURATION

inventory:
      "ansible_connection": "winrm",
      "ansible_winrm_transport": "http",
      "ansible_ssh_host": "mywinhost",
      "ansible_ssh_user": "myuser",
      "ansible_ssh_pass": "mypass",
      "ansible_ssh_port": 11111
##### OS / ENVIRONMENT

Controller: Centos 6.x
##### SUMMARY

If I don't use the default HTTP (non-ssl) port: ansible tries to use SSL connection regardless the "ansible_winrm_transport: http" setting.
##### STEPS TO REPRODUCE

Set up Winrm on a Win machine using default HTTP port: 5985
try to connect with "ansible_ssh_port: 5985"
then change the the port to another like 11111

```
winrm set winrm/config/Listener?Address=*+Transport=HTTP @{Port="11111"}
```

change inventory: "ansible_ssh_port: 11111"
Try to connect: SSL23_GET_SERVER_HELLO:unknown protocol
##### EXPECTED RESULTS

Expected: Ansible try to connect through HTTP (not https/ssl) and succeed
##### ACTUAL RESULTS

```
ansible mywinhost -m win_ping -o -vvvv
<mywinhost> ESTABLISH WINRM CONNECTION FOR USER: myuser on PORT 11111 TO mywinhost
<mywinhost> WINRM CONNECT: transport=plaintext endpoint=https://mywinhost:11111/wsman
<mywinhost> WINRM CONNECTION ERROR: 500 WinRMTransport. [Errno 1] _ssl.c:492: error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol
mywinhost | FAILED => 500 WinRMTransport. [Errno 1] _ssl.c:492: error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol
```

**Comment by bence-takacs** [comment]
UPDATE:
tried with "ansible_winrm_scheme: http" with basically the same result:

```
ansible mywinhost -m win_ping -o -vvvv
<mywinhost> ESTABLISH WINRM CONNECTION FOR USER: ci on PORT 11111 TO mywinhost
<mywinhost> WINRM CONNECT: transport=plaintext endpoint=https://mywinhost:11111/wsman
<mywinhost> WINRM CONNECTION ERROR: 500 WinRMTransport. [Errno 1] _ssl.c:492: error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol
mywinhost | FAILED => 500 WinRMTransport. [Errno 1] _ssl.c:492: error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol
```

The endpoint is still **https**://mywinhost:11111/wsman

**Comment by bence-takacs** [comment]
The issue is with the following file: /usr/lib/python2.6/site-packages/ansible/runner/connection_plugins/winrm.py

```
    def _winrm_connect(self):
        '''
        Establish a WinRM connection over HTTP/HTTPS.
        '''
        port = self.port or 5986
        vvv("ESTABLISH WINRM CONNECTION FOR USER: %s on PORT %s TO %s" % \
            (self.user, port, self.host), host=self.host)
        netloc = '%s:%d' % (self.host, port)
        exc = None
        for transport, scheme in self.transport_schemes['http' if port == 5985 else 'https']:
            if transport == 'kerberos' and (not HAVE_KERBEROS or not '@' in self.user):
                continue
            if transport == 'kerberos':
                realm = self.user.split('@', 1)[1].strip() or None
            else:
                realm = None
            endpoint = urlparse.urlunsplit((scheme, netloc, '/wsman', '', ''))
```

This does the magic :-(

> 'http' if port == 5985 else 'https'

Is this intentional, or a bug?
I see nothing about that behavior here: https://github.com/ansible/ansible/blob/stable-1.9/docsite/rst/intro_windows.rst

**Comment by nitzmahone** [comment]
If you must stay on 1.9 for some reason and do Windows stuff, at least use 1.9.5 (w/ pywinrm 0.2.0)- it has all the pywinrm arg passthru stuff you need to make this stuff work as-documented. Better still, get on 2.x for much better performance and something we'll actually support. :)

**Comment by bence-takacs** [comment]
Ok, then I'm going to upgrade to 1.9.5 but:
What do you mean by 'make this stuff work as-documented'? :-)

Do you mean that after installing ansible 1.9.5 (w/ pywinrm 0.2.0) the "ansible_winrm_scheme: http" will automatically work?

**Comment by jhawkesworth** [comment]
I think that is the case, yes @bence-takacs .  Historically ansible's documentation has only ever reflected the latest release or current development version, so the closer to the latest release you can get the more likely the documentation will be correct.  I would recommend getting to 2.0 or later if you can - there are many more windows modules available for one thing.

**Comment by bence-takacs** [comment]
Switching to 2.1 would mean regression testing all our playbooks against different systems and under different conditions.
According to the semantic versioning principle 1.9.5 would contain only (backward compatible) hotfixes compared to 1.9.1 - and the update should be safe :-)

**Comment by nitzmahone** [comment]
Just be aware that 1.9 is in a mostly-EOL mode, so if you run into anything that's not a major security flaw, it's unlikely to be supported/fixed, as you're running 2 "major" versions behind.

**Comment by bence-takacs** [comment]
After upgrading to 2.1 I'm getting this:
[WARNING]: ansible_winrm_scheme unsupported by pywinrm (is an up-to-date version of pywinrm installed?)

:-)

**Comment by jhawkesworth** [comment]
What version of pywinrm are you running?  

Are you running from source code?  You might need to run a 'make clean' to get rid of any old .pyc files following the upgrade.

Hope this helps,

Jon

**Comment by nitzmahone** [comment]
That's actually a bug introduced in 2.1.0 around pywinrm arg validation (the warning is incorrect and safe to ignore, it should still do what it's supposed to). It's the same issue as #16380 (but scheme instead of path)- already fixed for 2.1.1RC2 (which _should_ probably drop this week).

**Comment by bence-takacs** [comment]
Thanks for the clarification

All-in-all it seems to be working fine
